### ISSUE TITLE
Environment.SpecialFolder should expose FOLDERID_Downloads and FOLDERID_Public values

### ISSUE THREAD

**Issue by soc** [dotnet/runtime#25577]
`Environment.SpecialFolder` is the rough .NET equivalent of the [Known Folder](https://msdn.microsoft.com/en-us/library/windows/desktop/bb776911(v=vs.85).aspx) API, but it is lacking the corresponding values of `FOLDERID_Downloads` and `FOLDERID_Public`.

## Rationale

Other known folders like documents, pictures, videos, etc. are exposed, but those two (among others) are missing.

This seems to create confusion and misguided attempts [1](https://stackoverflow.com/questions/10667012/getting-downloads-folder-in-c) [2](https://stackoverflow.com/questions/3795023/how-to-programmatically-derive-windows-downloads-folder-userprofile-downloads) [3](https://social.msdn.microsoft.com/Forums/sqlserver/en-US/b344b1dd-d751-4e0e-9025-a391edd1949a/how-do-i-determine-the-windows-download-folder-path-using-the-knownfolders-class?forum=vbgeneral) to compute the paths to these folders by taking an existing value from `Environment.SpecialFolder`, dropping its last directory and appending "Downloads" to it. This is very error-prone, as the folder has not only different names based on the user locale, but also can be manually changed by users.

## Changes

Two values are added to the `SpecialFolder` enum:

```diff
          MyComputer = SpecialFolderValues.CSIDL_DRIVES,
+         Downloads = SpecialFolderValues.CSIDL_DOWNLOADS,
          MyMusic = SpecialFolderValues.CSIDL_MYMUSIC,
          MyPictures = SpecialFolderValues.CSIDL_MYPICTURES,
          MyVideos = SpecialFolderValues.CSIDL_MYVIDEO,
+         Public = SpecialFolderValues.CSIDL_PUBLIC,
          Recent = SpecialFolderValues.CSIDL_RECENT,
```

If this is desired, I'm happy to provide a pull request.

**Comment by Gnbrkm41** [comment]
Was almost writing a duplicate issue for this... copying some of the texts I've wrote there:
> The Downloads Folder isn't something that you can easily get the correct path to, in current versions of .NET. Most people would assume that `%USERPROFILE%/Downloads` would work, [however this can easily be configured to be placed elsewhere](https://www.tenforums.com/tutorials/74949-move-location-downloads-folder-windows-10-a.html) and because Windows doesn't place any symlinks to the actual download folder this could result in weird failures or undesirable behaviour of apps creating the download folder and ignoring user preferences.


I'm not sure how xplat this API has to be - but at the same a fair amount of the known folders are Windows specific (and we seem to have a lot of those Win-specific APIs in .NET Core already), so I'm not sure if xplat-ness is something we could ignore for this particular API.

With that said, [do we want all of the known folders](https://docs.microsoft.com/en-us/windows/win32/shell/knownfolderid) added to .NET's `SpecialFolder` (except ones marked as internal/deprecated)? There appears to be a fair amount of those that were introduced in Vista/7/8/10 and I'm not sure how many of them are missing.

I can write up a list tomorrow.

**Comment by danmosemsft** [comment]
We probably don't want all of them - just ones we believe are interesting/used. I think those that don't have Unix counterparts are less interesting, but as you say, we may need to put up with that.

**Comment by Gnbrkm41** [comment]
Here's the rough list of KNOWNFOLDERIDs (extracted from KnownFolders.h) and it's availability in .NET:


[summary.xlsx](https://github.com/dotnet/runtime/files/5197496/summary.xlsx)


KNOWNFOLDERID | .NET Equivalent | Deprecated/Internal | Version Introduced | [Listed on the docs?](https://docs.microsoft.com/en-us/windows/win32/shell/knownfolderid)
-- | -- | -- | -- | --
FOLDERID_AccountPictures |   |   | 8 | y
FOLDERID_AddNewPrograms |   |   |   | y
FOLDERID_AdminTools | AdminTools |   |   | y
FOLDERID_AllAppMods |   |   |   | n
FOLDERID_AppCaptures |   |   |   | n
FOLDERID_AppDataDesktop |   | Internal to .NET | 10 1709 | y
FOLDERID_AppDataDocuments |   | Internal to .NET | 10 1709 | y
FOLDERID_AppDataFavorites |   | Internal to .NET | 10 1709 | y
FOLDERID_AppDataProgramData |   | Internal to .NET | 10 1709 | y
FOLDERID_ApplicationShortcuts |   |   | 8 | y
FOLDERID_AppsFolder |   |   | 8 | y
FOLDERID_AppUpdates |   |   | Vista | y
FOLDERID_CameraRoll |   |   | 8.1 | y
FOLDERID_CameraRollLibrary |   |   |   | n
FOLDERID_CDBurning | CDBurning |   |   | y
FOLDERID_ChangeRemovePrograms |   |   |   | y
FOLDERID_CommonAdminTools | CommonAdminTools |   |   | y
FOLDERID_CommonOEMLinks | CommonOemLinks |   |   | y
FOLDERID_CommonPrograms | CommonPrograms |   |   | y
FOLDERID_CommonStartMenu | CommonStartMenu |   |   | y
FOLDERID_CommonStartMenuPlaces |   |   |   | n
FOLDERID_CommonStartup | CommonStartup |   |   | y
FOLDERID_CommonTemplates | CommonTemplates |   |   | y
FOLDERID_ComputerFolder | MyComputer |   |   | y
FOLDERID_ConflictFolder |   |   | Vista | y
FOLDERID_ConnectionsFolder |   |   |   | y
FOLDERID_Contacts |   |   | Vista | y
FOLDERID_ControlPanelFolder |   |   |   | y
FOLDERID_Cookies | Cookies |   |   | y
FOLDERID_CurrentAppMods |   |   |   | n
FOLDERID_Desktop | Desktop/DesktopDirectory |   |   | y
FOLDERID_DevelopmentFiles |   |   |   | n
FOLDERID_Device |   |   |   | n
FOLDERID_DeviceMetadataStore |   |   | 7 | y
FOLDERID_Documents | MyDocuments/Personal |   |   | y
FOLDERID_DocumentsLibrary |   |   | 7 | y
FOLDERID_Downloads |   |   |   | y
FOLDERID_Favorites | Favorites |   |   | y
FOLDERID_Fonts | Fonts |   |   | y
FOLDERID_Games |   | Deprecated (W10 1803) |   | y
FOLDERID_GameTasks |   |   | Vista | y
FOLDERID_History | History |   |   | y
FOLDERID_HomeGroup |   |   | 7 | y
FOLDERID_HomeGroupCurrentUser |   |   | 8 | y
FOLDERID_ImplicitAppShortcuts |   |   | 7 | y
FOLDERID_InternetCache | InternetCache |   |   | y
FOLDERID_InternetFolder |   |   |   | y
FOLDERID_Libraries |   |   | 7 | y
FOLDERID_Links |   |   |   | y
FOLDERID_LocalAppData | LocalApplicationData |   |   | y
FOLDERID_LocalAppDataLow |   |   |   | y
FOLDERID_LocalDocuments |   |   |   | n
FOLDERID_LocalDownloads |   |   |   | n
FOLDERID_LocalizedResourcesDir | LocalizedResources |   |   | y
FOLDERID_LocalMusic |   |   |   | n
FOLDERID_LocalPictures |   |   |   | n
FOLDERID_LocalVideos |   |   |   | n
FOLDERID_Music | MyMusic |   |   | y
FOLDERID_MusicLibrary |   |   | 7 | y
FOLDERID_NetHood | NetworkShortcuts |   |   | y
FOLDERID_NetworkFolder |   |   |   | y
FOLDERID_Objects3D |   |   | 10 1703 | y
FOLDERID_OneDrive |   |   |   | n
FOLDERID_OriginalImages |   |   | Vista | y
FOLDERID_PhotoAlbums |   |   | Vista | y
FOLDERID_Pictures | Pictures |   |   | y
FOLDERID_PicturesLibrary |   |   | 7 | y
FOLDERID_Playlists |   |   |   | y
FOLDERID_PrintersFolder |   |   |   | y
FOLDERID_PrintHood | PrinterShortcuts |   |   | y
FOLDERID_Profile | UserProfile |   |   | y
FOLDERID_ProgramData | CommonApplicationData |   |   | y
FOLDERID_ProgramFiles | ProgramFiles |   |   | y
FOLDERID_ProgramFilesCommon | CommonProgramFiles |   |   | y
FOLDERID_ProgramFilesCommonX64 | (Not directly exposed) |   |   | y
FOLDERID_ProgramFilesCommonX86 | CommonProgramFilesX86 |   |   | y
FOLDERID_ProgramFilesX64 | (Not directly exposed) |   |   | y
FOLDERID_ProgramFilesX86 | ProgramFilesX86 |   |   | y
FOLDERID_Programs | Programs |   |   | y
FOLDERID_Public |   |   | Vista | y
FOLDERID_PublicDesktop | CommonDesktopDirectory |   |   | y
FOLDERID_PublicDocuments | CommonDocuments |   |   | y
FOLDERID_PublicDownloads |   |   | Vista | y
FOLDERID_PublicGameTasks |   |   | Vista | y
FOLDERID_PublicLibraries |   |   | 7 | y
FOLDERID_PublicMusic | CommonMusic |   |   | y
FOLDERID_PublicPictures | CommonPictures |   |   | y
FOLDERID_PublicRingtones |   |   | 7 | y
FOLDERID_PublicUserTiles |   |   | 8 | y
FOLDERID_PublicVideos | CommonVideos |   |   | y
FOLDERID_QuickLaunch |   |   |   | y
FOLDERID_Recent | Recent |   |   | y
FOLDERID_RecordedCalls |   |   |   | n
FOLDERID_RecordedTV |   | Deprecated/Removed (7) |   | y
FOLDERID_RecordedTVLibrary |   |   | 7 | y
FOLDERID_RecycleBinFolder |   |   |   | y
FOLDERID_ResourceDir | Resources |   |   | y
FOLDERID_RetailDemo |   |   |   | n
FOLDERID_Ringtones |   |   | 7 | y
FOLDERID_RoamedTileImages |   |   | 8 | y
FOLDERID_RoamingAppData | ApplicationData |   |   | y
FOLDERID_RoamingTiles |   |   | 8 | y
FOLDERID_SampleMusic |   |   |   | y
FOLDERID_SamplePictures |   |   |   | y
FOLDERID_SamplePlaylists |   |   | Vista | y
FOLDERID_SampleVideos |   |   |   | y
FOLDERID_SavedGames |   |   | Vista | y
FOLDERID_SavedPictures |   |   |   | y
FOLDERID_SavedPicturesLibrary |   |   |   | y
FOLDERID_SavedSearches |   |   |   | y
FOLDERID_Screenshots |   |   | 8 | y
FOLDERID_SEARCH_CSC |   |   |   | y
FOLDERID_SEARCH_MAPI |   |   |   | y
FOLDERID_SearchHistory |   |   | 8.1 | y
FOLDERID_SearchHome |   |   |   | y
FOLDERID_SearchTemplates |   |   | 8.1 | y
FOLDERID_SendTo | SendTo |   |   | y
FOLDERID_SidebarDefaultParts |   |   | 7 | y
FOLDERID_SidebarParts |   |   | 7 | y
FOLDERID_SkyDrive |   | (Yes?) | 8.1 | y
FOLDERID_SkyDriveCameraRoll |   |   | 8.1 | y
FOLDERID_SkyDriveDocuments |   |   | 8.1 | y
FOLDERID_SkyDriveMusic |   |   |   | n
FOLDERID_SkyDrivePictures |   |   | 8.1 | y
FOLDERID_StartMenu | StartMenu |   |   | y
FOLDERID_StartMenuAllPrograms |   |   |   | n
FOLDERID_Startup | Startup |   |   | y
FOLDERID_SyncManagerFolder |   |   | Vista | y
FOLDERID_SyncResultsFolder |   |   | Vista | y
FOLDERID_SyncSetupFolder |   |   | Vista | y
FOLDERID_System | System |   |   | y
FOLDERID_SystemX86 | SystemX86 |   |   | y
FOLDERID_TreeProperties |   | Unsupported (7) |   | y
FOLDERID_Templates | Templates |   |   | y
FOLDERID_UserPinned |   |   | 7 | y
FOLDERID_UserProfiles |   |   | Vista | y
FOLDERID_UserProgramFiles |   |   | 7 | y
FOLDERID_UserProgramFilesCommon |   |   | 7 | y
FOLDERID_UsersFiles |   |   |   | y
FOLDERID_UsersLibraries |   |   | 7 | y
FOLDERID_Videos | MyVideos |   |   | y
FOLDERID_VideosLibrary |   |   | 7 | y
FOLDERID_Windows | Windows |   |   | y

Some notable ones are OneDrive/SkyDrive ones. There's separate entries in the header however OneDrive isn't listed on the docs. At the same time, the comments on the header says the SkyDrive one is deprecated and has the same ID as OneDrive's.

**Comment by soc** [comment]
@Gnbrkm41 Here is a cross-platform mapping people seem to have been happy with: https://github.com/dirs-dev/dirs-rs#features

**Comment by RayKoopa** [comment]
It should be noted that `CSIDL_DOWNLOADS` and `CSIDL_PUBLIC` do not actually exist as implied by the sample code in the proposal.

Digging through related issues, I found valid points rejecting extending the `SpecialFolder` enum; I tried to summarize the reasons in the newer discussion at https://github.com/dotnet/runtime/issues/554#issuecomment-1107831878

Despite good information collected here which should be retained, I think this discussion is stale(r) and should be closed in favor of [#554](https://github.com/dotnet/runtime/issues/554) as it covers this issue.

**Comment by danmoseley** [comment]
Closing as duplicate.
